### PR TITLE
[UIMA-6162] Concurrent binary serialization produces corrupt output

### PR DESCRIPTION
**Pull request by reckart**
- TUnit test which triggers the concurrent serialization data corruption situation

### CODE REVIEW

File: uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import org.apache.uima.cas.text.AnnotationFS;
+import org.apache.uima.resource.metadata.TypeSystemDescription;
+import org.junit.Test;
+
+public class ConcurrentBinarySerializationTest
+{
+    /**
+     * Serialization of the CAS is not inherently thread-safe. This test tries to run multiple
+     * serializations of the CAS in parallel to trigger a situation where an invalid serialization
+     * is generated - to be found by deserializing again.
```
Comment (mischor) [inline uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java]: I'm confused.  After the "synch" additions, serialization of the CAS should be thread-safe for multiple concurrent serializations.  But serialization of the CAS while another thread is updating the CAS (adding/modifying Feature Structures or adding/removing to/from indexes) is not safe (and is not currently "checked").

I assume that's what you meant to say?  This test should fail (before the sync fixes were added) and succeed after them?

Comment (reckart) [reply uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java]: Right - it should fail right now as the PR is based on commit 1fb339310e7a11cae5398f0cb8a2ec2eb636905e which doesn't include your synchronization yet. But when merging in your fixes from master, it should work. What surprises me is that the build doesn't seem to have failed...

Comment (reckart) [reply uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java]: Ah, Jenkins seems to temporarily merge the PR into master before building:

```
Commit message: "Merge e3cd9827a1424caeda481588f836d3a64e63816b into 382b096ae3967af6fa6fd6758fd36ae427372c6a"
```

Comment (reckart) [reply uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java]: Anyway - if you are good with the test, we could merge it.

Comment (mischor) [reply uimaj-core/src/test/java/org/apache/uima/cas/impl/ConcurrentBinarySerializationTest.java]: Great, Thanks for this very nice test case :-) .   I wonder if it works in uima v2?

**Pull request merged**